### PR TITLE
John conroy/react details default

### PR DESCRIPTION
**Pull request by john-conroy**
Towards #202 . I'll wait to close the issue until the visualizations can be filled in.

### CODE REVIEW

File: context/app/static/js/App.jsx [hunk 1]
```diff
@@ -40,3 +41,14 @@ export default function App(props) {
...
+App.propTypes = {
+  flaskData: PropTypes.exact({
+    entity: PropTypes.object,
+    flashed_messages: PropTypes.array,
+    provenance: PropTypes.object,
+    vitessce_conf: PropTypes.object,
+  }).isRequired,
+};
+
+export default App;
```
Comment (mccalluc) [inline context/app/static/js/App.jsx:54]: Is this a better practice than exporting at the definition?

File: context/app/static/js/components/Details.jsx [hunk 2]
```diff
@@ -69,3 +57,13 @@ export default function Details(props) {
     </Container>
   );
 }
+
+Details.propTypes = {
+  /* eslint-disable react/forbid-prop-types */
+  assayMetaData: PropTypes.object.isRequired,
+  provData: PropTypes.object.isRequired,
+  vitData: PropTypes.object.isRequired,
+  /* eslint-disable react/forbid-prop-types */
```
Comment (mccalluc) [inline context/app/static/js/components/Details.jsx:66]:
```suggestion
  /* eslint-enable */
```

File: context/app/static/js/components/RecursiveList.jsx [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                  {!isRoot && <PropertyName> {propertyName}: </PropertyName>}
+                  {Object.values(property).map((objProperty, index) => (
+                    <RecursiveList
+                  // eslint-disable-next-line react/no-array-index-key
+                      key={index}
+                      property={objProperty}
+                      propertyName={Object.getOwnPropertyNames(property)[index]}
+                      isRootChild={isRoot}
+                    />
+                  )) }
```
Comment (mccalluc) [inline context/app/static/js/components/RecursiveList.jsx:63]: This almost matches the block above... Would it be possible to factor it out and define it up at the top of the function, and then reference it here, either bare, or wrapped in `<DataPanel>`?

File: context/app/static/js/components/RecursiveList.jsx [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+function RecursiveList(props) {
+  const {
+    property, propertyName, isRoot, isRootChild,
+  } = props;
+
+
+  return (
+    /* eslint-disable no-nested-ternary */
```
Comment (mccalluc) [inline context/app/static/js/components/RecursiveList.jsx:29]: It is getting a little hard to read. Consider handling the `!property` and `isLeaf(property)` cases as guard clauses: That would take off two layers of nesting.

File: context/app/static/js/components/RecursiveList.jsx [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  font-size: 16px;
+`;
+
+const PropertyName = styled.span`
+  font-weight: bolder;
+  font-size: 1rem;
+`;
+
+function isLeaf(property) {
+  return (typeof property === 'number' || typeof property === 'string' || typeof property === 'boolean' || isEmptyArrayOrObject(property));
```
Comment (mccalluc) [inline context/app/static/js/components/RecursiveList.jsx:19]: consider using a set?
```suggestion
  return ( new Set(['number', 'string', 'boolean']).has(typeof property) || isEmptyArrayOrObject(property));
```

File: context/app/static/js/components/RecursiveList.jsx [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+              isRootChild={isRootChild}
+            />
+          ) : (
+            isRootChild
+              ? (
+                <DataPanel propertyName={propertyName} isRootChild={isRootChild}>
+                  {Object.values(property).map((objProperty, index) => (
+                    <RecursiveList
+                    // eslint-disable-next-line react/no-array-index-key
+                      key={index}
```
Comment (mccalluc) [inline context/app/static/js/components/RecursiveList.jsx:46]: Would this be any better?
```suggestion
                      key={objProperty}
```

File: context/app/static/js/helpers/functions.js [hunk 7]
```diff
@@ -0,0 +1,9 @@
+export function isEmptyArrayOrObject(val) {
+  let ret = false;
+  if (val.constructor.name === 'Object') {
+    ret = Object.keys(val).length === 0;
+  } else if (val.constructor.name === 'Array') {
+    ret = val.length === 0;
+  }
+  return ret;
```
Comment (mccalluc) [inline context/app/static/js/helpers/functions.js:8]: Avoid `let` and just return as soon as possible?
```suggestion
  if (val.constructor.name === 'Object') {
    return Object.keys(val).length === 0;
  }
  if (val.constructor.name === 'Array') {
    return val.length === 0;
  }
  return false;
```

File: context/app/static/js/App.jsx (hunk 1 above)
Comment (john-conroy) [reply context/app/static/js/App.jsx:54]: More of a preference than a better practice. Not strongly opinionated on this one so if you believe I should revert the change where possible, I can.

File: context/app/static/js/components/RecursiveList.jsx (hunk 6 above)
Comment (john-conroy) [reply context/app/static/js/components/RecursiveList.jsx:46]: The key would then be one of an object, array, number or string. Could change the key to the 'propertyName' ?

File: context/app/static/js/App.jsx (hunk 1 above)
Comment (mccalluc) [reply context/app/static/js/App.jsx:54]: No, I don't care one way or the other, just curious. Just leave it. Thanks!

**Comment by john-conroy** [comment]
Resolved the issues @mccalluc mentioned. UI inside the expansion panels isn't great by any means, but I am weary of putting in much more time into it since it's a stop gap before the design is finished.

**Comment by mccalluc** [comment]
@tsliaw : Can you look at this? Unless you find something abhorrent, I'd suggest that this be merged, and you take it as a starting point for iterations on the design.

**Comment by john-conroy** [comment]
Thanks! I think those issues are best handled separately.

**Pull request merged**